### PR TITLE
Remove polymorphization

### PR DESCRIPTION
**Pull request by saethlin**
This PR removes the flag `-Zpolymorphize` and all the infrastructure in the compiler that exists only to support it, per https://github.com/rust-lang/compiler-team/issues/810.

### CODE REVIEW

File: tests/ui/coroutine/polymorphize-args.rs [hunk 1]
```diff
@@ -1,4 +1,3 @@
-//@ compile-flags: -Zpolymorphize=on
```
Comment (compiler-errors) [inline tests/ui/coroutine/polymorphize-args.rs:1]: pls remove this test, it's no longer needed if we're not polymorphizing anything.

File: tests/ui/mir/inline-wrong-abi.rs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-//@ compile-flags: -Zpolymorphize=on -Zinline-mir=yes -Zmir-opt-level=0
```
Comment (compiler-errors) [inline tests/ui/mir/inline-wrong-abi.rs:1]: Can u change this to `-Zvalidate-mir`? The presence of `-Zpolymorphize` this test was just to observe (and thus compute and validate) the optimized MIR before codegen.

File: tests/ui/unsized/issue-115809.rs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-//@ compile-flags: --emit=link -Zmir-opt-level=2 -Zpolymorphize=on
```
Comment (compiler-errors) [inline tests/ui/unsized/issue-115809.rs:1]: pls remove or replace with `-Zvalidate-mir`

File: compiler/rustc_ty_utils/src/abi.rs [hunk 4]
```diff
@@ -45,23 +45,8 @@ fn fn_sig_for_fn_abi<'tcx>(
...
-            // (i.e. due to being inside a projection that got normalized, see
-            // `tests/ui/polymorphization/normalized_sig_types.rs`), and codegen not keeping
-            // track of a polymorphization `ParamEnv` to allow normalizing later.
-            //
-            // We normalize the `fn_sig` again after instantiating at a later point.
             let mut sig = match *ty.kind() {
-                ty::FnDef(def_id, args) => tcx
-                    .fn_sig(def_id)
-                    .map_bound(|fn_sig| {
-                        tcx.normalize_erasing_regions(
```
Comment (compiler-errors) [inline compiler/rustc_ty_utils/src/abi.rs:59]: yippee

File: compiler/rustc_monomorphize/src/partitioning.rs [hunk 5]
```diff
@@ -661,9 +661,7 @@ fn characteristic_def_id_of_mono_item<'tcx>(
                     return None;
                 }
 
-                // When polymorphization is enabled, methods which do not depend on their generic
-                // parameters, but the self-type of their impl block do will fail to normalize.
-                if !tcx.sess.opts.unstable_opts.polymorphize || !instance.has_param() {
```
Comment (compiler-errors) [inline compiler/rustc_monomorphize/src/partitioning.rs:666]: isn't this now unconditionally true? `!false || WHATEVER` == `true`?

File: tests/ui/mir/inline-wrong-abi.rs (hunk 2 above)
Comment (saethlin) [reply tests/ui/mir/inline-wrong-abi.rs:1]: I really wasn't sure what of the affected tests should be salvaged. Thanks for the guidance ❤️

**Comment by saethlin** [comment]
r? compiler-errors

**Comment by compiler-errors** [comment]
r=me after it's green

@bors rollup=never 

(since it likely has a net-positive perf result, but let's not block this PR on perf.)

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
looks like you've got an unnecessary import for `TypeVisitableExt`

File: compiler/rustc_codegen_cranelift/build_system/tests.rs [hunk 6]
```diff
@@ -92,10 +92,6 @@ const BASE_SYSROOT_SUITE: &[TestCase] = &[
     TestCase::build_bin_and_run("aot.mod_bench", "example/mod_bench.rs", &[]),
     TestCase::build_bin_and_run("aot.issue-72793", "example/issue-72793.rs", &[]),
     TestCase::build_bin("aot.issue-59326", "example/issue-59326.rs"),
-    TestCase::custom("aot.polymorphize_coroutine", &|runner| {
-        runner.run_rustc(&["example/polymorphize_coroutine.rs", "-Zpolymorphize"]);
```
Comment (bjorn3) [inline compiler/rustc_codegen_cranelift/build_system/tests.rs:96]: This test file can be removed too.

**Comment by saethlin** [comment]
@compiler-errors I'm not 100% sure I rebased properly over the change in `compiler/rustc_ty_utils/src/abi.rs`.

I also expanded my search a bit and found a few more mentions of polymorphization to delete.

**Comment by compiler-errors** [comment]
lgtm, r=me

**Comment by lqd** [comment]
Massively positive results, regressions are small and in stress tests only, marking as triaged.

@rustbot label: +perf-regression-triaged

**Comment by nnethercote** [comment]
Is it possible to briefly summarize the source of the performance improvements? I'm just curious.

**Comment by saethlin** [comment]
This PR deletes an often-called query and its associated table. I believe most of the improvement is caused by having less nodes in the query dep graph. (which I would call query overhead)

**Comment by nnethercote** [comment]
Was it an oversight that the query was being called despite not doing anything useful? Or was it more complicated than that?

**Comment by bjorn3** [comment]
The result of the unused_generic_params query was previously encoded in the crate metadata in case a crate that depends on the local crate uses `-Zpolymorphize`.

**Comment by compiler-errors** [comment]
I feel like the code perhaps couldve been written to not touch the polymorphization codepaths unless the feature was enabled, though not totally certain given polymorphization's interaction with codegen 🤔 

not certain if the unused parameters query couldve been pushed onto the crates that actually needed it (ie the ones build with Zpolymorphize)

**Pull request merged**